x/authenticator/authenticator/composite.go

```diff
@@ -0,0 +1,140 @@
+package authenticator
+
+import (
+	"encoding/json"
+	"strconv"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/osmosis-labs/osmosis/v23/x/authenticator/iface"
+)
+
+type SubAuthenticatorInitData struct {
+	AuthenticatorType string `json:"authenticator_type"`
+	Data              []byte `json:"data"`
+}
+
+func subTrack(
+	ctx sdk.Context,
+	account sdk.AccAddress, msg sdk.Msg, msgIndex uint64,
+	authenticatorId string,
+	subAuthenticators []iface.Authenticator,
+) error {
+	for id, auth := range subAuthenticators {
+		err := auth.Track(ctx, account, msg, msgIndex, compositeId(authenticatorId, id))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type PassingReq int
+
+const (
+	requireAllPass = iota
+	requireAnyPass
+)
+
+func subHandleRequest(
+	ctx sdk.Context,
+	request iface.AuthenticationRequest,
+	subAuthenticators []iface.Authenticator,
+	passingReq PassingReq,
+	f func(auth iface.Authenticator, ctx sdk.Context, request iface.AuthenticationRequest) error,
+) error {
+	if passingReq != requireAllPass && passingReq != requireAnyPass {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "invalid passing req")
+	}
+
+	var err error
+
+	for id, auth := range subAuthenticators {
+		// update the authenticator id to include the sub-authenticator id
+		request.AuthenticatorId = compositeId(request.AuthenticatorId, id)
+
+		err = f(auth, ctx, request)
+
+		if passingReq == requireAllPass && err != nil {
+			return err
+		}
+
+		if passingReq == requireAnyPass && err == nil {
+			return nil
+		}
+	}
+
+	// require all pass return no error if it has not yet early returned
+	if passingReq == requireAllPass {
+		return nil
+	}
+
+	// require any pass return error it has not yet early returned
+	if passingReq == requireAnyPass {
+		return err
```

Comment:
what error are we returning here? It seems like the last one. I think we want either all the errors or a generic error saying all failed (in which case the other errors should be logged and/or emitted as events)